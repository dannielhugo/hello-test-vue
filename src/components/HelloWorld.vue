<template>
  <div class="hello">
    <ul>
      <li v-for="link in links" :key="link.text">
        <a :href="link.href" target="_blank" rel="noopener">{{link.text}}</a>
      </li>
    </ul>

    <button @click="fetchStarships">FETCH STARSHIPS</button>

    <ul v-if="starships.length > 0" class="starships">
      <li v-for="starship in starships" :key="starship.name">
        <p>Model: {{starship.model}}</p>
        <p>Name: {{starship.name}}</p>
        <p>Credits: {{starship.cost_in_credits}}</p>
      </li>
    </ul>
  </div>
</template>

<script>
import axios from 'axios';

export default {
  name: 'HelloWorld',
  props: {
    // SNAPSHOT TEST
    // showAwesomeVue: {
    //   type: Boolean,
    //   default: true,
    // },
  },
  data() {
    return {
      links: [
        {href: 'https://router.vuejs.org', text: 'vue-router'},
        {href: 'https://vuex.vuejs.org', text: 'vuex'},
        {href: 'https://github.com/vuejs/vue-devtools#vue-devtools', text: 'vue-devtools'},
        {href: 'https://vue-loader.vuejs.org', text: 'vue-loader'},
      ],
      starships: []
    };
  },
  async mounted() {
    // SNAPSHOT TEST
    // if (this.showAwesomeVue) {
    //   this.links.push(
    //     {href: 'https://github.com/vuejs/awesome-vue', text: 'awesome-vue'},
    //   );
    // }
  },
  methods: {
    async fetchStarships() {
      const response = await axios.get('https://swapi.dev/api/starships/');
      this.starships = response.data.results;
    }
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
h3 {
  margin: 40px 0 0;
}
ul {
  list-style-type: none;
  padding: 0;
}
li {
  display: inline-block;
  margin: 0 10px;
}
a {
  color: #42b983;
}

.starships {
  display: flex;
  flex-direction: column;
}

.starships li {
  border: 2px solid #f1f1f1;
  border-radius: 1.5rem;
  margin-top: 1rem;
}
</style>
