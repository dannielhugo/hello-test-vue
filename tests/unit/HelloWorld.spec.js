import { shallowMount } from '@vue/test-utils'
import HelloWorld from '@/components/HelloWorld.vue'

import moxios from 'moxios';
import flushPromises from 'flush-promises';

const STARSHIPS_MOCK= [
  { name: 'T-65 X-wing', model: 'X-wing', cost_in_credits: 149999},
  { name: 'T-65 X-wing 2', model: 'X-wing', cost_in_credits: 169999}
]

describe('HelloWorld.vue', () => {
  let wrapper;
  beforeEach(() => {
    moxios.install();
    wrapper = shallowMount(HelloWorld, {});
  })

  afterEach(() => {
    moxios.install();
  })

  it('should render vue links', () => {
      expect(wrapper.html()).toMatchSnapshot();
  })

  it('should have no starships', async () => {
    const starshipList = wrapper.find('.starships');

    expect(starshipList.exists()).toBeFalsy();
  })

  it('should fetch starships on button click', async () => {
    moxios.stubRequest('https://swapi.dev/api/starships/', {
      status: 200,
      response: {
        results: STARSHIPS_MOCK
      }
    });

    wrapper.find('button').trigger('click')
    await moxios.wait();
    await flushPromises();

    const starshipList = wrapper.find('.starships');

    expect(starshipList.isVisible()).toBeTruthy();

    const starships = starshipList.findAll('li');
    expect(starships.length).toBe(2);
  })
})
