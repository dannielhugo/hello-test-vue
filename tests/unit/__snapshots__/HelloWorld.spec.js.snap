// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelloWorld.vue should render vue links 1`] = `
<div class="hello">
  <ul>
    <li><a href="https://router.vuejs.org" target="_blank" rel="noopener">vue-router</a></li>
    <li><a href="https://vuex.vuejs.org" target="_blank" rel="noopener">vuex</a></li>
    <li><a href="https://github.com/vuejs/vue-devtools#vue-devtools" target="_blank" rel="noopener">vue-devtools</a></li>
    <li><a href="https://vue-loader.vuejs.org" target="_blank" rel="noopener">vue-loader</a></li>
  </ul> <button>FETCH STARSHIPS</button>
  <!---->
</div>
`;
